test/powershell/engine/Api/MultithreadedEngine.Tests.ps1

```diff
@@ -0,0 +1,198 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+Describe 'Multithreaded engine APIs' -Tags 'CI' {
+    BeforeAll {
+        $sbStub = @'
+.foreach{
+    [pscustomobject]@{
+        Time       = [DateTime]::Now.ToString('yyyyMMddTHHmmss.fffffff')
+        Value      = $_
+        ThreadId   = [System.Threading.Thread]::CurrentThread.ManagedThreadId
+        RunspaceId = [runspace]::DefaultRunspace.Id
+    }
+    Start-Sleep -Milliseconds 500
+}
+'@
+    }
+
+    Context 'PowerShell::InvokeAsync - Single script tests' {
+        BeforeAll {
+            function InvokeAsyncHelper {
+                param(
+                    [powershell]$PowerShell,
+                    [switch]$Wait = $false
+                )
+                $sb = [scriptblock]::Create("@(1,2,3,4,5)${sbStub}")
+                $r = $PowerShell.AddScript($sb).InvokeAsync()
+                if ($Wait) {
+                    [System.Threading.Tasks.Task]::WaitAll(@($r))
+                }
+                $r
+            }
+
+            function GetInnerErrorId {
+                param(
+                    [exception]$Exception
+                )
+                while ($null -ne $Exception.InnerException) {
+                    $Exception = $Exception.InnerException
+                    if ($null -ne (Get-Member -InputObject $Exception -Name ErrorRecord)) {
+                        $Exception.ErrorRecord.FullyQualifiedErrorId
+                        break
+                    }
+                }
+            }
+        }
+
+        It 'can invoke a single script asynchronously, but only in a new runspace' {
+            $ps = [powershell]::Create()
+            $r = InvokeAsyncHelper -PowerShell $ps -Wait
+            $r.Status | Should -Be ([System.Threading.Tasks.TaskStatus]::RanToCompletion)
+            $r.IsCompletedSuccessfully | Should -Be $true
+        }
+
+        It 'handles terminating errors properly when invoked asynchronously in a new runspace' {
+            $ps = [powershell]::Create()
+            $sb = {
+                $r = $ps.AddScript(@'
+try {
+    Get-Process -InvalidParameter 42 -ErrorAction Stop
+} catch {
+    throw
+}
```

Comment:
That's necessary for the terminating error to actually be terminating. Otherwise the terminating error is treated as non-terminating. It's a really annoying PowerShell issue that I want to fix in another PR eventually.

Reply:
Ah-ha, I was looking for this! Can you please provide me an example of this annoying behavior? For some context, see @lzybkr's comment here: https://github.com/PowerShell/PowerShell/pull/8425#issuecomment-467564258

I have been trying to figure out why the try/catch/throw was put there in `mkdir` and `oss`.

Reply:
Even though this isn't in the scope of this PR, just to flush this out, since you asked I wanted to comment: PowerShell has a very annoying behavior where terminating errors are not terminating unless they are wrapped in try/catch. That means as a module author if I don't take the time to wrap each begin, process and end block in functions exported by my module in try/catch, then I risk users of my module seeing different behavior depending on whether or not they invoke my functions inside of try/catch. i.e. I could have a function that treats a certain error as benign and keep running after the error and then have users who download a module containing that function see different behavior if they invoke it in a try/catch. The only way for PowerShell script/function authors to ensure that the consumers of those scripts/functions get the same behavior every time is to use a top-level try/catch to force terminating errors to actually terminate.

In the case of your `mkdir` and `oss` functions, it may not be necessary; however, to ensure you get consistent error handling it is a best practice that many script authors use because of the incorrect behavior in PowerShell (terminating errors not terminating unless wrapped in try/catch). The right solution is to not have to wrap it in try/catch at all, and simply fix PowerShell to that terminating errors always terminate.